### PR TITLE
Re-use same ContiguousArray for tasks copy operations.

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We allocated a new ContiguousArray per each EventLoop run when tasks needed to be executed. We can just re-use the same instance all the time and so safe malloc operations (we will need to do memset tho).
Beside this we also did not reserve the needed capacity up fron which could lead to realloc operations.

Modifications:

- Reuse same ContiguousArray instance
- Correctly set the initial capacity

Result:

Less overhead when processing tasks via the EventLoop

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Lukasa huh? We use it (just no in the defer method)

**Comment by Lukasa** [comment]
Ah, sorry, I missed the one line that made that change! 🤦‍♂️

File: Sources/NIO/EventLoop.swift [hunk 1]
```diff
@@ -325,6 +327,8 @@ internal final class SelectableEventLoop: EventLoop {
         self._addresses = UnsafeMutablePointer.allocate(capacity: Socket.writevLimitIOVectors)
         self.msgs = UnsafeMutableBufferPointer(start: _msgs, count: Socket.writevLimitIOVectors)
         self.addresses = UnsafeMutableBufferPointer(start: _addresses, count: Socket.writevLimitIOVectors)
+        // We will process 4096 tasks per while loop.
+        self.tasksCopy.reserveCapacity(4096)
```
Comment (Lukasa) [inline Sources/NIO/EventLoop.swift]: Is 4096 really a good choice here?

Comment (normanmaurer) [reply Sources/NIO/EventLoop.swift]: I honestly have no idea.... I just used something here which did not seems to high and so waste too much memory. Suggestions ?

File: Sources/NIO/EventLoop.swift [hunk 2]
```diff
@@ -502,13 +507,12 @@ internal final class SelectableEventLoop: EventLoop {
...
                     break
                 }
-                var tasksCopy = ContiguousArray<() -> Void>()
 
                 // We only fetch the time one time as this may be expensive and is generally good enough as if we miss anything we will just do a non-blocking select again anyway.
                 let now = DispatchTime.now()
 
                 // Make a copy of the tasks so we can execute these while not holding the lock anymore
-                while let task = scheduledTasks.peek(), task.readyIn(now) <= .nanoseconds(0) {
+                while tasksCopy.count < tasksCopy.capacity, let task = scheduledTasks.peek(), task.readyIn(now) <= .nanoseconds(0) {
```
Comment (normanmaurer) [inline Sources/NIO/EventLoop.swift]: @Lukasa @weissi note that I also ensure here we never need to grow it.

File: Sources/NIO/EventLoop.swift [hunk 3]
```diff
@@ -464,16 +468,17 @@ internal final class SelectableEventLoop: EventLoop {
     public func run() throws {
         precondition(self.inEventLoop, "tried to run the EventLoop on the wrong thread.")
         defer {
-            var tasksCopy = ContiguousArray<ScheduledTask>()
-
+            var scheduledTasksCopy = ContiguousArray<ScheduledTask>()
             tasksLock.lock()
+            // reserve the correct capacity so we don't need to realloc later on.
+            scheduledTasksCopy.reserveCapacity(scheduledTasks.count)
```
Comment (weissi) [inline Sources/NIO/EventLoop.swift]: shouldn't this happen before we take the lock?

Comment (normanmaurer) [reply Sources/NIO/EventLoop.swift]: We need the lock because we access scheduledTasks.count

Comment (weissi) [reply Sources/NIO/EventLoop.swift]: thanks, makes sense

**Pull request merged**